nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/jwt/JwtAuthenticationFilter.java

```diff
@@ -43,33 +51,63 @@ public Authentication attemptAuthentication(final HttpServletRequest request) {
             return null;
         }
 
-        // TODO: Refactor request header extraction logic to shared utility as it is duplicated in AccessResource
+        // Check for JWT in cookie and header
+        final String cookieToken = getTokenFromCookie(request);
+        final String headerToken = getTokenFromHeader(request);
+        if (cookieToken != null && !cookieToken.isEmpty()) {
+            if (!UNAUTHENTICATED_METHODS.contains(request.getMethod().toUpperCase())) {
+                // To protect against CSRF when using a cookie, if the request method requires authentication the request must have a matching Authorization header JWT
+                if (headerToken.equals(cookieToken)) {
+                    return new JwtAuthenticationRequestToken(headerToken, request.getRemoteAddr());
+                } else {
+                    throw new InvalidAuthenticationException("Authorization HTTP header and authentication cookie did not match.");
+                }
+            } else {
+                return new JwtAuthenticationRequestToken(cookieToken, request.getRemoteAddr());
+            }
+        } else if (headerToken != null && !headerToken.isEmpty()) {
```

Comment:
It would be helpful to refactor this set of conditionals to make the logic easier to follow.  Part of the challenge here seems to be attempting to accomplish multiple goals in a single method.  It would be easier to follow if CSRF protection were handled separate from JWT request token retrieval, even though they are related.

Leveraging Spring Security's CsrfFilter and customizing token retrieval and comparison seems like a better way to enforce CSRF protection.  The CsrfFilter default configuration already handled skipping idempotent HTTP methods, so it would be a better of customizing the CsrfTokenRepository implementation to retrieve the appropriate header, and then compare it against the same value provided in the `Authorization` header.

Separating the implementation will require some more, but it should help avoid confusion, and avoid potential pitfalls by using less custom code.

Reply:
I have refactored this by adding the NiFiCsrfTokenRepository which will check the Cookie header for a 'CSRF' token that should match the JWT Authorization header. This simplifies the JwtAuthenticationFilter and pushed the idempotent methods checks to the Spring CsrfFilter code.